cluster-autoscaler/proposals/resourceslack.md

```diff
@@ -0,0 +1,266 @@
+* [R1] Works on relatively large clusters today's CA used to work
+  * dozens of nodes, hundreds of pods
+  * In general we don't want to have O(pods * nodes) operations every single loop
+  * [R1-1] On large cluster there is a huge number of pods to be scheduled. And the pods are mostly the same. If a pod from one replica set X failed to schedule on node A then there is no point to in trying to fit the next pod from X on A. Thus there should be just a single loop over all nodes (ref https://github.com/kubernetes/autoscaler/pull/56#discussion_r115942312)
+  * [R1-2] The scale down performance strongly depends that the pods don't change every iteration. So if you placed placeholder-X-123 on node A then in the next iteration it should also land on A or the scale down performance will be heavily impacted (ref https://github.com/kubernetes/autoscaler/pull/56#discussion_r115942312)
+* [R2] Keep the specified amount of extra capacity at minimum
+  * If we allowed users to annotate their k8s pods so that they are mirrored to equivalent placeholder pods like https://github.com/kubernetes/autoscaler/pull/77#discussion_r117084488,
+    CA may actually produce much more extra capacity than specified by the user.
+
+    Let's say an user specified CA to reserve 10% extra capacity, but also marked a pod which requests cpu and memory more than 10% of current capacity. CA will result in creating a placeholder pod which consumes resource more than 10%.
+    So, the "10%" in this case is just a minimum requirement. CA may reserve more if the user instructed CA to do so by annotations. That's why I wrote `at minimum`. It is impossible to specify CA an exact rate of extra capacity when we want to suppoprt [R6] and [R7].
+* [R3] Ensure full compatibility with scale down
+* [R4] Only expose necessary and sufficient knobs to users
+
+Version 2
+
+contains [R1]-[R4] and also what is below:
+
+* [R5] Don't add unnecessarily large node
+  * A node is unnecessarily large when:
+    * Any pod that is intended to be scheduled faster by resource slack fit within a smaller node
+* [R6] Possible to provide extra space even for the "biggest" pod in term of memory and cpu
+  * However, if the biggest pod in the cluster was a database, we won't want to keep a hot standby node for the pod to move to
+* [R7] Possible to provide extra space even for every pod which is not biggest but do request some cpu(< max(pod_request_cpu)) and memory(< max(pod_request_mmory))
+  * Say, the "biggest pods" request 1000m cpu/1G mem, 10m cpu/10G mem respectively
+  * Even if they could fit within existing free space, a 500m(<1000m cpu)/5G mem pod has chances not to fit any node 
+* [R8] Reserve space even for the pods having anti pod affinity
+* [R9] Reserve space even for the pods having node selector/node affinity
+
+Version 3
+
+contains [R1]-[R9] and also what is below:
+
+* [R10] Works on considerably large clusters
+  * To move CA from beta to GA we need to handle 5000 nodes and 150 000 pods.
```

Comment:
@MaciekPytel 

> It leaves the question on how we choose the size of those placeholders. A reasonable default placeholder size with a flag to override it?

I generally agree to that 👍
I'm unsure though if we could end up with a "reasonable" default because I guess we don't know what could be really "reasonable" across not all but many clusters.

> Some function that determines the size of placeholders based on something?

How about moving it to version 2+ of this feature, so that we can proceed anyway?

> my instinct would be to pick something very simple for initial version and update later if required.

I agree 👍